drivers/spi/spi_silabs_eusart.c

```diff
@@ -227,54 +693,114 @@ static int spi_silabs_eusart_init(const struct device *dev)
+	err = clock_control_on(eusart_config->clock_dev, 
+							(clock_control_subsys_t)&eusart_config->clock_cfg);
 	if (err < 0) {
 		return err;
 	}
 
+	err = pinctrl_apply_state(eusart_config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0 && (err != -ENOENT)) {
+		/** The return number ENOENT means that no sleep state is defined by the user in the
+		 *  DT.
+		 */
+		return err;
+	}
+
 	err = spi_context_cs_configure_all(&data->ctx);
 	if (err < 0) {
 		return err;
 	}
+
 	spi_context_unlock_unconditionally(&data->ctx);
 
-	return 0;
+	pm_device_init_off(dev);
+
+	err = pm_device_runtime_enable(dev);
+
+	return err;
 }
 
 static int spi_silabs_eusart_transceive(const struct device *dev, const struct spi_config *config,
-				       const struct spi_buf_set *tx_bufs,
-				       const struct spi_buf_set *rx_bufs)
+					const struct spi_buf_set *tx_bufs,
+					const struct spi_buf_set *rx_bufs)
 {
 	struct spi_silabs_eusart_data *data = dev->data;
+	const struct spi_silabs_eusart_config *eusart_config = dev->config;
 	uint16_t control = 0;
 	int ret;
 
 	spi_context_lock(&data->ctx, false, NULL, NULL, config);
 
 	ret = spi_silabs_eusart_configure(dev, config, &control);
-	if (ret < 0) {
+
+	if (ret) {
 		spi_context_release(&data->ctx, ret);
 		return ret;
 	}
 
 	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
-	spi_silabs_eusart_xfer(dev, config);
+#if defined(CONFIG_SPI_SILABS_EUSART_DMA)
+	if (eusart_config->is_dma_enabled_instance) {
+		ret = spi_silabs_eusart_xfer_dma(dev, config);
+	} else
+#endif
+	{
+		spi_silabs_eusart_xfer_polling(dev, config);
+	}
 
 	spi_context_release(&data->ctx, ret);
 
-	return 0;
+	return ret;
 }
 
-#ifdef CONFIG_SPI_ASYNC
+#if defined(CONFIG_SPI_ASYNC) 
 static int spi_silabs_eusart_transceive_async(const struct device *dev,
-					     const struct spi_config *config,
-					     const struct spi_buf_set *tx_bufs,
-					     const struct spi_buf_set *rx_bufs,
-					     struct k_poll_signal *async)
+					      const struct spi_config *config,
+					      const struct spi_buf_set *tx_bufs,
+					      const struct spi_buf_set *rx_bufs, spi_callback_t cb,
+					      void *userdata)
 {
-	return -ENOTSUP;
+	struct spi_silabs_eusart_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+
+	uint16_t control = 0;
+	int ret = 0U;
+
+	/* Validate if the alignment of the buffers are valid. */
+	spi_context_lock(ctx, true, cb, userdata, config);
+
+	ret = spi_silabs_eusart_configure(dev, config, &control);
+
+	if (ret) {
+		spi_context_release(ctx, ret);
+		return ret;
+	}
+
+	spi_context_buffers_setup(ctx, tx_bufs, rx_bufs, 1);
+#if defined(CONFIG_SPI_SILABS_EUSART_DMA)
+	struct spi_silabs_eusart_config *eusart_config = dev->data;
+
+	if (eusart_config->is_dma_enabled_instance) {
+		ret = spi_silabs_eusart_xfer_dma(dev, config);
+	} else
+#endif
+	{
+		/** Async without dma is not implmented, please configure the device tree
+		 *  instance with the proper dma configuration.
+		 */
+		return -ENOTSUP;
```

Comment:
Context is not released.

BTW, locate all the clean-up code in a unique place:

```
	if (ret) {
		goto out;
	}
[...]
out:
	spi_context_release(&data->ctx, ret);
	return ret;
```